### PR TITLE
Fix the problem that the metastore-client is not closed

### PR DESCRIPTION
**Pull request by southernriver**
The hiveClienPool is not closed in unit test, Here we can add close() function to fix memory leak in all unit test as follow :
https://github.com/apache/iceberg/pull/2785/commits/5d8de6f79270665b6fb068717d80155a22e6ed0f#diff-f6a9c582e0cbe50c837dc3c5bd6b7e6db17527977180242bc3c5dd8df668cf28R70. 

Another solution is that we need to add this close() into shutdownhook since  it can solve all situation especially user's program.

### CODE REVIEW

File: baseline.gradle [hunk 1]
```diff
@@ -79,6 +79,7 @@ subprojects {
           '-Xep:StrictUnusedVariable:OFF',
           '-Xep:TypeParameterShadowing:OFF',
           '-Xep:TypeParameterUnusedInFormals:OFF',
+          '-Xep:ShutdownHook:OFF',
```
Comment (RussellSpitzer) [inline baseline.gradle:82]: Does this disable JVM Shutdown Hooks? I am a little confused that we are disabling this when we add a shutdown hook?

Comment (southernriver) [reply baseline.gradle:82]: It just avoid building failure for gradle.
As  https://github.com/palantir/gradle-baseline mentioned:
`ShutdownHook: Applications should not use Runtime#addShutdownHook.
`
and we can add this properties to skip this.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java [hunk 2]
```diff
@@ -65,6 +68,14 @@ private synchronized void init() {
               .removalListener((key, value, cause) -> ((HiveClientPool) value).close())
               .build();
     }
+    Runtime.getRuntime().addShutdownHook(new Thread() {
+      @Override
+      public void run() {
+        LOG.info("Before clientPool closed, clientPool().getCurrentSize() = {}", clientPool().getCurrentSize());
```
Comment (kbendick) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java:74]: Nit: Would this be better as a lower log level? Additionally, would it make more sense to format this as human readable / formatted text as opposed to references to functions (which might be awkward for end users)?

For example: `The CachedClientPool has a size of {}`.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java [hunk 3]
```diff
@@ -65,6 +68,14 @@ private synchronized void init() {
...
+        clientPool().close();
```
Comment (kbendick) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java:75]: Do we need to check if `close` has already been called and/or if the result of calling `clientPool()` is not null?

File: core/src/main/java/org/apache/iceberg/ClientPoolImpl.java [hunk 4]
```diff
@@ -140,4 +140,7 @@ private void release(C client) {
   public int poolSize() {
     return poolSize;
   }
+  public int getCurrentSize() {
```
Comment (pvary) [inline core/src/main/java/org/apache/iceberg/ClientPoolImpl.java:143]: nit: newline

File: hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java [hunk 5]
```diff
@@ -65,6 +68,16 @@ private synchronized void init() {
               .removalListener((key, value, cause) -> ((HiveClientPool) value).close())
               .build();
     }
+    Runtime.getRuntime().addShutdownHook(new Thread() {
```
Comment (pvary) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/CachedClientPool.java:71]: nit: newline

**Review by southernriver — COMMENTED** [review]
address comment

**Pull request closed without merging**